Validations refactor (RIPD-1412)

bachase opened the pull request:
This PR refactors the `Validations` class into a generic version that can be adapted for the consensus simulation framework.  `RCLValidations` replaces the current implementation by adapting the generic version to use RCL validations.  

The first two changesets are minor refactorings that simplify the generic implementation.
The final changeset cherry-picks @scottschurr's work in #2083 and could use his review.

bachase commented:
Documentation preview http://bachase.github.io/ripd1412/index.html

scottschurr commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -240,24 +240,45 @@ RCLValidations::onStale(RCLValidation&& v)
+        ScopedLockType sl(lock_);
 
-    Validations::flush();
+        Validations::flush([&](RCLValidation && v)
```
Wince.  When possible I try not to call outside functions with a `lock` held.  This code base doesn't hold to that rule very well, so I'll leave this for your consideration.  But at this point the only place we need the `ScopedLockType ` is for the `staleValidations_.emplace_back(std::move(v.val_)); ` inside the upcoming lambda.

Consider putting a lock inside the lambda (yes, the lock will thrash more) and postpone this `ScopedLockType` to just before the `if (anyNew && !writing_) ` a few lines down.

bachase replied:
If I move the lock inside, do I need to consider concurrent calls to `RCLValidations::flush`?  My understanding is that the lock protects _all_ data members of the class, both in the old and new code.  If another call to `RCLValidations::flush` occurs, wouldn't this cause a data race?

As long as I document the semantics of `flush`, I'm not too concerned about calling it while holding the lock and would prefer to avoid thrashing.  If we are comfortable calling the external code in `.emplace_back` while holding a lock, I think extending that comfort to the `flush` function is acceptable.   Does that seem reasonable or would you prefer I move the lock inside?

scottschurr replied:
You may be looking at the code more closely than I am, but I'm not seeing any data races if the lock is inside the lambda.  `anyNew` is a local, so it is not subject to thread races.

If we put the lock inside the lambda, we also need to add a lock just in front of line 268 to protect the rest of the function.  If that's what you're worried about then I agree.  We need a lock there as well.

In terms of what is "outside code" my personal approach is to have these buckets:
- `std` library code.
- non-`std` library, but well vetted, library code (like boost and beast).
- code in this same class.
- everything else (including other functions/classes in this code base).

Most calls into `std` library code I don't worry about.  The characteristics of that code are stable and well understood.  I get more nervous as I go further down the list because the code is less well characterized and more volatile.  Note that, in my opinion, calling into our own code outside this class is the scariest.  It is the most subject to change and we have the smallest understanding of the consequences of these changes.

Let me provide some more background for my concern.  When I hold a lock and then call an outside (non-`std` library) function I am, in effect, saying two things:
- It is okay if I hold this lock indefinitely, since I don't know how long this lock will be held.
- I strongly believe that there will be no lock order problems introduced by the outside code trying to acquire another lock or to re-acquire this lock.

I believe that you've checked all of the code that you're modifying at this moment.  I'm sure it's all good right now.

My primary concern is that, as the outside code is maintained, those maintainers have no idea that this lock is being held.  So problems like the ones I just described are easy to introduce during maintenance.

As I said, the existing code base is pretty bad (in my opinion) about holding locks over external calls.  And sometimes holding a lock over an external call is just flat-out unavoidable.  So you'll not be doing anything that isn't seen elsewhere in the code base.  I suspect in this case it *is* avoidable (at the cost of some extra lock thrash).  I'm just pointing it out for your consideration.

You know best how you should manage the lock, and I leave that to your discretion.  I'm sure you'll make the best choice for the code base.  Thanks for listening.

scottschurr commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -240,24 +240,45 @@ RCLValidations::onStale(RCLValidation&& v)
         return;
```
I'm worried about `RCLValidations::onStale()`.  In the old implementation the moral equivalent of `onStale()` was always called with `RCLValidations::lock_` held by a `std::lock_guard`.  Now all of the calls to `onStale` are coming from outside `RCLValidations`.  So I'm fairly confident the callers don't hold the `lock_`.

Any data members of `RCLValidations` that are accessed by `RCLValidations::doWrite()` need to always be protected by `lock_`.  That would include both `staleValidations_` and `writing_`.

It may be sufficient to simply grab the `lock_` at the very top of `onStale()`.  I'm unsure because I don't know the requirements of all of the places that call `onStale()`.  The result might look like this:
```
    {
        ScopedLockType sl(lock_);
        staleValidations_.emplace_back(std::move(v.val_));
        if (writing_)
            return;

        writing_ = true;
    }
    app_.getJobQueue().addJob(...
```
Once the locking situation is cleared up I'll look hard at the code again.

bachase replied:
The intent is that `onStale` is only accessible and callable from the generic `Validations` code and any calls into the `Validations` code from `RCLConsensus` require holding the lock for the duration of the call to the base class.  The fact that this is opaque is a shortcoming of the CRTP design.  I'll try to think of a way to make it clearer outside of just commenting.  I'm open to suggestions too =).

JoelKatz commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -0,0 +1,337 @@
+        return true;
+    }
+
+    // FIXME: This never forwards untrusted validations
```
@scottschurr The idea was that we would have a certain number of validation slots with priority going to validators we trusted. Remaining slots might be allocated to validators that were listed by publishers we trusted but that we didn't choose to trust. The shorter term plan was just to forward untrusted validations if peers wanted them or if we had the ability/bandwidth to. None of that was implemented.

JoelKatz commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,504 @@
+        LedgerID const& currentLedger,
+        LedgerID const& priorLedger,
+        std::size_t cutoffBefore,
+        beast::Journal& j)
```
It's supposedly cheap to copy, but it's hard to see how that could be as cheap as not copying.

wilsonianb replied:
If we got rid of the one trace log in this function, Validations.h wouldn't need to include `<ripple/beast/utility/Journal.h>`

bachase replied:
I was hoping to eliminate as well, but this one seems useful.  I'm about done with switching to a policy based design and I ended up having `Validations` own the journal even though it rarely uses it.

bachase commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -0,0 +1,337 @@
+    std::vector<NetClock::time_point> times;
+    {
+        ScopedLockType sl(lock_);
+        if (auto map = byLedger(ledger))
```
This is an important point and I agree. I tend to overuse auto when developing intially while the types are in flux. But readability for the future is essential. I will go back and fix.

seelabs commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,504 @@
+
+        hash_map<LedgerID, ValidationCounts> ret;
+
+        current(t, [&](auto const&, auto const& v) {
```
I use `auto` params in lambdas for two reasons:

1) When I need the param to be a template
2) When I want a terse syntax

When I don't need a generic lambda, I would apply the same rule of thumb for auto parameters as I do for auto variables. For example, if the lambdas is used near to where it is defined, and the parameter types are easily found at the nearby caller, `auto` is acceptable.

In this case, you can make a strong argument for declaring the parameter types. However, I don't want a rule that says "if a lambda doesn't have to be generic don't use `auto` params".

scottschurr replied:
I'm not advocating for a rule.  However I'd like to point out another example in the code base that I was just struggling with:

https://github.com/ripple/rippled/blob/develop/src/ripple/rpc/impl/ServerHandlerImp.cpp#L363-L365

The declaration looks like this:
```
    m_jobQueue.postCoro(jtCLIENT, "WS-Client",
        [this, session = std::move(session),
            jv = std::move(jv)](auto const& c)
        {
        ...
```
My question to you is this: as a maintainer of this code, what is the type of `c`?  [Mickey doesn't get to answer 😄 ]?

It turns out that 'c' can *only* be one type: `std::shared_ptr<JobQueue::Coro> const&`.

The length of that type name might very well be past the threshold of someone looking for a terse syntax.  But, for someone reading the code, the information is difficult to extract by just looking at the sources.  I guess my point is that a long type name can look like line noise unless you need to know what the type is.  Then it changes from line noise into valuable information.

Again, this is just my personal take on the subject.  I'm not arguing for a rule or for a change to the coding guidelines.  Thanks for listening.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,504 @@
+        This is a safety to protect against very old validations and the time
+        it takes to adjust the close time accuracy window.
+    */
+    std::chrono::seconds VALIDATION_CURRENT_WALL = std::chrono::minutes{5};
```
These were previously in `LedgerTiming` and were `constexpr` there, so I this was just a copy and paste.  However, I want the ability to change them within the simulation framework so they are no longer statically constant.  Instead, the `Validations` class takes the `ValidationParms` in its constructor and stores it as a const member.  I  can switch to `validation_CURRENT_WALL` or `validationCurrentWall` instead.  Do you prefer the latter?

scottschurr replied:
I think it depends on how you want folks to think about the values.  It sounds like the only reason they would change is for the simulation framework?  So do you want readers to think of these as though they are constants or as non-constants?  I suspect they are more likely to think of them as constants with names like`validationCURRENT_WALL`.  They are more likely to think of them as ordinary values with names like `validationCurrentWall`.  Naming conventions are mostly good for communicating intent to the reader/maintainer.  I guess I have a mild leaning toward `validationCurrentWall `, but I don't really have enough context for my opinion to carry much weight.

bachase replied:
I'd prefer most readers to view them as constant since only the simulation code should consider changing them at runtime.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,504 @@
+        : byLedger_(c), parms_(p)
+    {
+    }
+
```
1. I intend `Validations` to be used as a base class via CRTP only and prefer to not have any virtual table if possible (yes premature optimization!).  I have to handle different types, so virtuals aren't much help.  Is there a technique to indicate that aside from a comment?

2. No, it must be derived from.  It think it would fail to compile otherwise due to the `static_cast` in the `impl()` helper.

I had considered avoiding CRTP altogether and just taking in a policy class that implements the `onStale` callback desired by callers.  Would that be preferable?

scottschurr replied:
I don't know any technique to indicate that even though this is a base class it doesn't need a virtual destructor (other than a comment).  That doesn't mean there isn't such a technique...

We all prefer what we're accustomed to 😄.  I've used policy classes a lot, and they feel comfortable to me.  However other folks on the team have spent more time with the CRTP pattern and that feels comfortable to them.

My personal inclination is to avoid inheritance when it's not needed for run-time polymorphism.  That plus the fact that you specifically do not want a virtual table suggests to me that inheritance is the wrong answer here.  But that's my own bias.  I think you get to make your own call here.

bachase commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -0,0 +1,611 @@
+            return s;
+        }
+
+        std::size_t s;
```
This was just for the purpose of the unit test, which doesn't attempt to match the types of real validations/ledger sequences, but only conform to the general type requirements of a `Validation`, e.g. "has a `seq` that is copyable, comparable, `>=` and convert to `bool`.  My preference was to avoid relying on RCL-specific types.

That being said, there is no reason why I can't make it `std::uint32_t`!

bachase commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -0,0 +1,611 @@
+        operator std::size_t() const
+        {
+            return s;
+        }
```
I'll consider, but since this is test code, I'm inclined towards terseness in the testcases below.

bachase commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -0,0 +1,611 @@
+            hash_append(h, id.id);
+        }
+
+        std::size_t id;
```
I'm happy to change, but in principle this test should be independent of the bitwidth of `std::size_t`.  If a platform has a different width, its doing an extra test for us! ;)

bachase commented on src/ripple/app/misc/Validations.cpp:
```diff
@@ -46,38 +47,34 @@ class ValidationsImp : public Validations
     }
 
-    std::shared_ptr<ValidationSet> findSet (uint256 const& ledgerHash)
+    ValidationSet * findSet (uint256 const& ledgerHash)
```
Correct, this was temporary while I refactored.  In the new version of the code in `Validations`, the equivalent lookup into the aged map still returns a `ValidationSet *`, but no public methods can return that set so the lifetime is guaranteed.

As you might be thinking, that is only "enforced" by convention and a soon-to-be-added comment.  An alternative I considered was to completely restrict access to the map and only allow iteration over it via a functor.

bachase commented:
@scottschurr has brought up some good points that might warrant switching away from the CRTP design.  I'll take a day or two to see if it will stick and will potentially re-submit the review if it makes more sense.  Feel free to put this on the back-burner until I comment further.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,504 @@
+        AddOutcome result = AddOutcome::current;
+
+        // Attempt to insert
+        auto ins = current_.emplace(key, val);
```
If we need to replace the current validation because it is stale, we need to [modify `ins`](https://github.com/ripple/rippled/pull/2084/files/c4a607efb55e31bbcc66b5fa4ca91c68780b500f..03fbaadad201877c0c6ca13249d0d303517c1af5#diff-46f83c3cd15dd825b63068c0ab4dea00R369)

bachase commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -0,0 +1,337 @@
+        ScopedLockType sl(lock_);
+        auto res = Validations::add(now(), *pubKey, val);
+
+        // This is a duplicate validadtion
```
Dad joke?

wilsonianb commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -0,0 +1,611 @@
+        // first round a,b,c agree
+        for(auto const & node : {a,b,c})
+            BEAST_EXPECT(AddOutcome::current == vals.add(node, Seq{1}, ID{1}));
+
```
You could add a trusted node `d` with validations for different ledger ids.

wilsonianb commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -0,0 +1,611 @@
+        }
+
+        {
+            // Cutoff old sequence numberss
```
numberss --> numbers

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,506 @@
+
+                NodeID ni = v.nodeID();
+                if (ni > p.highNode)
+                    p.highNode = ni;
```
I'd like that too.  @JoelKatz is that acceptable?

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,506 @@
+        if a time is specified
+
+        @param t (Optional) Time used to determine staleness
+        @param f Callable with signature (NodeKey const &, Validations const &)
```
It had come up in the past, it seems to be a bug in how docca renders some templates.

bachase commented:
I just pushed two new changesets.  The first addresses many of the stylistic and `auto`-abuse issues @scottschurr identified.  The second switches from CRTP to a policy design and addresses issues @wilsonianb identified.  

I decided to have the generic `Validations` class own the instance of the policy as a private member.  I believe this helps indicate that the policy class is completely subservient to the `Validations` instance.  The `RCLValidationsPolicy` owns the stale validations and has its own mutex for managing access.  The `Validations` class has its own mutex for managing access to its members--the `current` and `byLedger` `Validation`s.  I've avoided calling into the policy instance while holding a lock, save for one [`onStale` call](https://github.com/ripple/rippled/pull/2084/files#diff-46f83c3cd15dd825b63068c0ab4dea00R257) which I believe is reasonable to avoid some copies.

I kept the history of commits for reviewers, but will gladly squash if that is easier.

bachase commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -0,0 +1,254 @@
+    bool shouldRelay = false;
+
+    // only add trusted or listed
+    if (val->isTrusted() || pubKey)
```
@wilsonianb This was my approach to eliminating the extra `isCurrent` call.  Can you spot check it?

bachase commented:
Updated documentation preview @ http://bachase.github.io/ripd1412/index.html

wilsonianb commented:
I'm getting these compile warnings with the new commits
```
In file included from src/test/consensus/Validations_test.cpp:22:0,
                 from src/test/unity/consensus_test_unity.cpp:22:
src/ripple/consensus/Validations.h: In instantiation of 'ripple::Validations<StalePolicy, Validation, MutexType>::Validations(const ripple::ValidationParms&, beast::abstract_clock<std::chrono::_V2::steady_clock>&, beast::Journal, Ts&& ...) [with Ts = {ripple::test::Validations_test::StaleData&, beast::manual_clock<std::chrono::_V2::steady_clock>&}; StalePolicy = ripple::test::Validations_test::StalePolicy; Validation = ripple::test::Validations_test::Validation; MutexType = ripple::test::Validations_test::DummyMutex]':
src/test/consensus/Validations_test.cpp:380:63:   required from here
src/ripple/consensus/Validations.h:227:20: warning: 'ripple::Validations<ripple::test::Validations_test::StalePolicy, ripple::test::Validations_test::Validation, ripple::test::Validations_test::DummyMutex>::j_' will be initialized after [-Wreorder]
     beast::Journal j_;
                    ^
src/ripple/consensus/Validations.h:225:17: warning:   'ripple::test::Validations_test::StalePolicy ripple::Validations<ripple::test::Validations_test::StalePolicy, ripple::test::Validations_test::Validation, ripple::test::Validations_test::DummyMutex>::stalePolicy_' [-Wreorder]
     StalePolicy stalePolicy_;
                 ^
src/ripple/consensus/Validations.h:303:5: warning:   when initialized here [-Wreorder]
     Validations(
     ^
```

scottschurr commented:
@bchase: Seeing this error on Travis build 9129.4:
```
src/test/consensus/Validations_test.cpp:892:34: error: no matching literal operator for call to 'operator""s' with argument of type 'unsigned long long' or 'const char *', and no matching literal operator template
        harness.clock().advance(1s);
                                 ^
1 error generated.
```
Are you missing a `using namespace std::chrono_literals;`?

scottschurr commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -36,264 +36,74 @@
-RCLValidations::doWrite(ScopedLockType&)
+RCLValidationsPolicy::doStaleWrite(ScopedLockType&)
 {
     std::string insVal(
```
`insVal` and `findSeq` could be `static const`

scottschurr commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -36,264 +36,74 @@
+            doStaleWrite(sl);
         }
 
         // Handle the case where flush() is called while a queuedWrite
```
`... while a doStaleWrite() is...`?

scottschurr commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -36,264 +36,74 @@
 }
 
-// NOTE: doWrite() must be called with mLock *locked*.  The passed
+// NOTE: doStaleWrite() must be called with mLock *locked*.  The passed
```
mLock -> staleLock_?

scottschurr commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -352,7 +162,93 @@ RCLValidations::doWrite(ScopedLockType&)
+    {
+        using AddOutcome = RCLValidations::AddOutcome;
+
+        AddOutcome res = validations.add(*pubKey, val);
```
`const res`?

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -323,23 +331,23 @@ class Validations
-            auto const oldSeq = oldVal.seq();
-            auto const newSeq = val.seq();
+            std::uint32_t const oldSeq = oldVal.seq();
+            std::uint32_t const newSeq = val.seq();
```
Thanks for catching this!  I intended to make seq `std::uint32_t` everywhere and had accidentally left it as a `std::size_t` in `RCLValidation`.

bachase commented on src/ripple/app/consensus/RCLValidations.h:
```diff
@@ -0,0 +1,207 @@
+        return val_->getLedgerHash();
+    }
+
+    /// Validated ledger's sequence number (0 if none)
```
I can make it more explicit, but I believe that `getFieldByValue` will already to that https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/STObject.h#L573

bachase replied:
I see I wasn't consistent with that for `loadFee`.  I'll add the explicit check.

bachase commented on src/ripple/app/consensus/RCLValidations.h:
```diff
@@ -0,0 +1,207 @@
+     /** Flush current validations to disk before shutdown.
+
+         @param remaining The remaining validations to flush
+      */
```
I'm having trouble spotting it on my end.  Which part is misaligned?

wilsonianb replied:
https://github.com/bachase/rippled/blob/2f0b7e02eb36dbcda73fed871d33f098483054d1/src/ripple/app/consensus/RCLValidations.h#L175-L179

bachase commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -0,0 +1,254 @@
+        }
+
+        // Handle the case where flush() is called while a queuedWrite
+        // is already in progress.
```
This was a port of the changes in #2083.  If `flush` is called while an existing async job was queued or is now running, this would block until it completes.  It would be nice to drop this if we don't care about whether that job finishes before returning.

@scottschurr: What is your opinion?  Is it important for the `flush` call to block until the potentially asynchronous call is complete?

scottschurr replied:
@wilsonianb and @bachase I think the loop is still useful.  This is a thread synchronization point.  A different thread may have called `doStaleWrite()`.  If that thread is running when we enter `flush()` at shutdown, then `staleWriting_` will be `true` (line 89).  So we'll skip calling `doStaleWrtie()` ourselves and we'll let the other thread do our job for us.  However when we exit this function all of the stale cache should be flushed to the database.  The loop makes sure we don't leave the function until the stale cache is all safely in the database.

I had hoped the comments were sufficient to describe what we're up to. but maybe not.  @wilsonianb, do you have a suggested change for the comments?

wilsonianb replied:
ok, that makes sense to me.
I'd vote for this being clearly conveyed in the comment:
>when we exit this function all of the stale cache should be flushed to the database. The loop makes sure we don't leave the function until the stale cache is all safely in the database.

bachase commented:
Rebased on 0.70.0-b4 and addressed the latest set of comments.  I extending the private `byLedger` and `current` members of `Validations` to allow reserving storage prior to iterating items.  This does so under the same lock used for iteration.  The interface is a tad more mucky, but since these are private helper members, it seems worth the trade-off. 

Thanks @wilsonianb and @scottschurr for your careful reviews; the code is definitely better for it.

bachase commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -199,11 +202,11 @@ handleNewValidation(Application& app,
     bool shouldRelay = false;
 
     // only add trusted or listed
```
Yes, still true.  `pubKey` is only set if the validation is trusted or listed.

bachase commented on src/ripple/app/misc/Validations.cpp:
```diff
@@ -1,544 +0,0 @@
-                    soci::transaction tr(*db);
-                    for (auto it: vector)
-                    {
-                        s.erase ();
```
Nice catch!!!!

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,690 @@
+                    maybeStaleValidation.emplace(std::move(oldVal));
+                    // Replace old val in the map and set the previous ledger ID
+                    ins.first->second = val;
+                    ins.first->second.setPreviousLedgerID(oldID);
```
That makes sense.  I'll cherry-pick.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -0,0 +1,690 @@
+            boost::none,
+            [&](std::size_t) {}, // nothing to reserve
+            [&](NodeKey const&, Validation const& v) {
+                if (v.trusted() && v.isPreviousLedgerID(ledgerID))
```
Good point.  I think that may be the best we can do though, since we have a gap in the validation stream from that validator?

wilsonianb replied:
One option would be for `getNodesAfter` and `currentTrustedDistribution` to take a ledger sequence instead of a `LedgerID`.
In both cases, we don't care if there was actually a validation for the previous ledger. We only care that a validation is for ledger that follows a particular ledger.

bachase replied:
Without the `LedgerID`, wouldn't we be counting ledgers that are potentially on a different ledger chain?  At least, for `currentTrustedDistrubution`, I think care about that distinction.

wilsonianb replied:
Yeah, that would change the assumption about what's being counted in `currentTrustedDistrubution`. The problem is that the count under the current assumption (same ledger chain) can be inaccurate if we don't set the previous ledger id for a validation. It seems like we have the information to determine that a lone validation validates a ledger id whose sequence is immediately after a given sequence. It would just be a matter of finding the least messy way of figuring that out.

bachase commented:
I will squash and rebase after the final 👍 .  I also tested running `rippled` for a few minutes and looking at the validations in the sqlite database.

wilsonianb commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -480,6 +486,14 @@ class Validations_test : public beast::unit_test::suite
                 BEAST_EXPECT(harness.vals().numTrustedForLedger(ID{20}) == 1);
+
+                // A new key, but re-issue the same validation, since it is the
+                // same, it is considered a repeat
```
I think we want this be `AddOutcome::current` for this case:
https://github.com/ripple/rippled/pull/2084#discussion_r113582740
We would need to change this check in `Validations::add`
https://github.com/ripple/rippled/pull/2084/files#diff-46f83c3cd15dd825b63068c0ab4dea00R383
to something like
```
auto const ret = byLedger_[id].emplace(key, val);
if (! ret.second && ret.first->second.key() == val.key())
    return AddOutcome::repeat;
```

bachase replied:
Good call.  I should've noticed that too.

wilsonianb replied:
Good test.

bachase replied:
FYI, there was an [additional change](https://github.com/bachase/rippled/blob/231984ba1a2e0ea47fce5fc0865827173f88f108/src/ripple/consensus/Validations.h#L420-L432) for this.

wilsonianb commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -498,8 +512,28 @@ class Validations_test : public beast::unit_test::suite
+
             }
+            {
+                // Processs validations out of order with shifted times
```
Processs --> Process

wilsonianb commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -498,8 +512,28 @@ class Validations_test : public beast::unit_test::suite
+                BEAST_EXPECT(
+                    AddOutcome::current == harness.add(a, Seq{8}, ID{8}));
+
+                // Process a validation that has "later" seq but early sign time
```
Interesting how sign time trumps ledger sequence, but it makes sense.

bachase commented:
Rebased and squashed after addressing last set of @wilsonianb's comments.

scottschurr commented:
Build error in CircleCI:
```
In file included from src/ripple/unity/app_consensus.cpp:21:
In file included from src/ripple/app/consensus/RCLConsensus.cpp:22:
In file included from src/ripple/app/consensus/RCLValidations.h:24:
src/ripple/consensus/Validations.h:422:47: error: no member named
      'insert_or_assign' in 'std::unordered_map<ripple::PublicKey,
      ripple::RCLValidation, beast::uhash<beast::spooky>,
      std::equal_to<ripple::PublicKey>, std::allocator<std::pair<const
      ripple::PublicKey, ripple::RCLValidation> > >'
                                validationMap.insert_or_assign(key, val);
                                ~~~~~~~~~~~~~ ^
src/ripple/app/consensus/RCLValidations.cpp:210:44: note: in instantiation of
      member function 'ripple::Validations<ripple::RCLValidationsPolicy,
      ripple::RCLValidation, std::mutex>::add' requested here
        AddOutcome const res = validations.add(*pubKey, val);
                                           ^
```

bachase commented:
Argg, C++17 support strikes again!

bachase commented:
Thanks @wilsonianb and @scottschurr for your thorough reviews.  This is a much better change now.

I'll consider the `ValidaitonssCounts/highNode` change in a separate PR.

bachase commented:
Squashed and rebased on 0.70.0-b6

bachase commented:
Rebased onto 0.70.0-b8 and added the changest from #2094 to avoid merge conflicts when putting into a future release.

bachase commented:
Rebased on 0.70.1

seelabs commented:
Merged into 0.80.0-b1
